api/build/create.go

```diff
@@ -0,0 +1,392 @@
+	// before compiling. After we're done compiling, we reset the pipeline type.
+	r.SetPipelineType(pipelineType)
+
+	// skip the build if only the init or clone steps are found
+	skip := SkipEmptyBuild(p)
+	if skip != "" {
+		// set build to successful status
+		input.SetStatus(constants.StatusSuccess)
+
+		// send API call to set the status on the commit
+		err = scm.FromContext(c).Status(u, input, r.GetOrg(), r.GetName())
+		if err != nil {
+			logger.Errorf("unable to set commit status for %s/%d: %v", r.GetFullName(), input.GetNumber(), err)
+		}
+
+		c.JSON(http.StatusOK, skip)
+
+		return
+	}
+
+	// check if the pipeline did not already exist in the database
+	//
+	//nolint:dupl // ignore duplicate code
+	if pipeline == nil {
+		pipeline = compiled
+		pipeline.SetRepoID(r.GetID())
+		pipeline.SetCommit(input.GetCommit())
+		pipeline.SetRef(input.GetRef())
+
+		// send API call to create the pipeline
+		err = database.FromContext(c).CreatePipeline(pipeline)
+		if err != nil {
+			retErr := fmt.Errorf("unable to create new build: failed to create pipeline for %s: %w", r.GetFullName(), err)
+
+			util.HandleError(c, http.StatusBadRequest, retErr)
+
+			return
+		}
+
+		// send API call to capture the created pipeline
+		pipeline, err = database.FromContext(c).GetPipelineForRepo(pipeline.GetCommit(), r)
+		if err != nil {
+			//nolint:lll // ignore long line length due to error message
+			retErr := fmt.Errorf("unable to create new build: failed to get new pipeline %s/%s: %w", r.GetFullName(), pipeline.GetCommit(), err)
+
+			util.HandleError(c, http.StatusInternalServerError, retErr)
+
+			return
+		}
+	}
+
+	input.SetPipelineID(pipeline.GetID())
+
+	// create the objects from the pipeline in the database
+	err = PlanBuild(database.FromContext(c), p, input, r)
+	if err != nil {
+		util.HandleError(c, http.StatusInternalServerError, err)
+
+		return
+	}
+
+	// send API call to update repo for ensuring counter is incremented
+	err = database.FromContext(c).UpdateRepo(r)
+	if err != nil {
+		retErr := fmt.Errorf("unable to create new build: failed to update repo %s: %w", r.GetFullName(), err)
+
+		util.HandleError(c, http.StatusBadRequest, retErr)
+
+		return
+	}
+
+	// send API call to capture the created build
+	input, _ = database.FromContext(c).GetBuildForRepo(r, input.GetNumber())
+
+	c.JSON(http.StatusCreated, input)
+
+	// send API call to set the status on the commit
+	err = scm.FromContext(c).Status(u, input, r.GetOrg(), r.GetName())
+	if err != nil {
+		logger.Errorf("unable to set commit status for build %s/%d: %v", r.GetFullName(), input.GetNumber(), err)
+	}
+
+	// publish the build to the queue
+	go PublishToQueue(
+		queue.FromGinContext(c),
+		database.FromContext(c),
+		p,
+		input,
+		r,
+		u,
+	)
+}
+
+// getPRNumberFromBuild is a helper function to
+// extract the pull request number from a Build.
+func getPRNumberFromBuild(b *library.Build) (int, error) {
```

Comment:
this is used for both `CreateBuild` and `UpdateBuild` but is in `create.go` - should it go in a file on its own? 🤔 

It's probably fine here, but I'm asking anyway. 🤷

Reply:
Going to leave it there for now, but I almost wonder if this belongs in types rather than as a helper function